mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -882,6 +882,132 @@ struct ForwardOpPattern : public ConvertOpToLLVMPattern<ForwardOp> {
+        SmallVector<Value> residuals;
+        auto params = op.getArguments();
+
+        for (auto i = 0; i < argc * 2; i++) {
+            bool isDup = (i % 2) != 0;
+            Value val = params[i];
+            isDup ? differentials.push_back(val) : inputs.push_back(val);
+        }
+
+        auto upperLimit = (argc * 2) + (resc * 2);
+        for (auto i = argc * 2; i < upperLimit; i++) {
+            bool isDup = (i % 2) != 0;
+            Value val = params[i];
+            isDup ? cotangents.push_back(val) : outputs.push_back(val);
+        }
+
+        auto tapeCount = op.getTape();
+        auto uppestLimit = upperLimit + tapeCount;
+        for (auto i = upperLimit; i < uppestLimit; i++) {
+            residuals.push_back(params[i]);
+        }
+
+        SmallVector<Type> newFuncInputTys;
+
+        for (auto [in, diff] : llvm::zip(inputs, differentials)) {
+            newFuncInputTys.push_back(in.getType());
+            newFuncInputTys.push_back(diff.getType());
+        }
+
+        for (auto [out, cotan] : llvm::zip(outputs, cotangents)) {
+            newFuncInputTys.push_back(out.getType());
+            newFuncInputTys.push_back(cotan.getType());
+        }
+
+        SmallVector<Type> residualStructs;
+        auto converter = getTypeConverter();
+        for (auto residualMemref : residuals) {
+            auto residualMemrefTy = residualMemref.getType();
+            auto residualStructTy = converter->convertType(residualMemrefTy);
+            residualStructs.push_back(residualStructTy);
+        }
+
+        auto ctx = rewriter.getContext();
+        // This gives me a struct of the following type
+        // { memref0, memref1, memref2, ... memrefN }
+        auto tapeTy = LLVM::LLVMStructType::getLiteral(ctx, residualStructs);
+        // This gives me { { memref, ..., memrefn } }
+        auto wrappedFlatTapeStructTy = LLVM::LLVMStructType::getLiteral(ctx, {tapeTy});
+        auto ptrTy = LLVM::LLVMPointerType::get(ctx);
+        Type retty = tapeCount > 0 ? dyn_cast<Type>(wrappedFlatTapeStructTy) : dyn_cast<Type>(ptrTy);
+        newFuncInputTys.push_back(retty);
+
+        auto newFuncTy = FunctionType::get(ctx, newFuncInputTys, TypeRange{});
+
+        ModuleOp mod = op->getParentOfType<ModuleOp>();
+        PatternRewriter::InsertionGuard guard(rewriter);
+        rewriter.setInsertionPointToStart(mod.getBody());
+
+        auto func = rewriter.create<mlir::func::FuncOp>(op.getLoc(), op.getSymName(), newFuncTy);
+        func.setPrivate();
+
+        Block *entry = func.addEntryBlock();
+        rewriter.setInsertionPointToStart(entry);
+
+        IRMapping map;
+
+        if (tapeCount > 0) {
+        auto loc = op.getLoc();
+        auto lastIdx = newFuncInputTys.size() - 1;
+        Value wrappedStructValAgg = func.getArgument(lastIdx);
+
+        SmallVector<Value> tapestructs;
+        for (auto i = 0; i < tapeCount; i++) {
+            SmallVector<int64_t> pos = {0, static_cast<int64_t>(i)};
```

Comment:
Maybe I don't need the zero here.

Reply:
Mmm... tried this patch:

```
diff --git a/mlir/lib/Gradient/Transforms/ConversionPatterns.cpp b/mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
index eb4802aa..c31bda2b 100644
--- a/mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
+++ b/mlir/lib/Gradient/Transforms/ConversionPatterns.cpp
@@ -941,10 +941,8 @@ struct ReverseOpPattern : public ConvertOpToLLVMPattern<ReverseOp> {
         // This gives me a struct of the following type
         // { memref0, memref1, memref2, ... memrefN }
         auto tapeTy = LLVM::LLVMStructType::getLiteral(ctx, tapeStructs);
-        // This gives me { { memref, ..., memrefn } }
-        auto wrappedFlatTapeStructTy = LLVM::LLVMStructType::getLiteral(ctx, {tapeTy});
         auto ptrTy = LLVM::LLVMPointerType::get(ctx);
-        Type inputTapeTy = tapeCount > 0 ? dyn_cast<Type>(wrappedFlatTapeStructTy) : dyn_cast<Type>(ptrTy);
+        Type inputTapeTy = tapeCount > 0 ? dyn_cast<Type>(tapeTy) : dyn_cast<Type>(ptrTy);
         newFuncInputTys.push_back(inputTapeTy);
 
         auto newFuncTy = FunctionType::get(ctx, newFuncInputTys, TypeRange{});
@@ -968,7 +966,7 @@ struct ReverseOpPattern : public ConvertOpToLLVMPattern<ReverseOp> {
 
         SmallVector<Value> tapestructs;
         for (auto i = 0; i < tapeCount; i++) {
-            SmallVector<int64_t> pos = {0, static_cast<int64_t>(i)};
+            SmallVector<int64_t> pos = {static_cast<int64_t>(i)};
             Value tapeStructIth =
                 rewriter.create<LLVM::ExtractValueOp>(loc, wrappedStructValAgg, pos);
             tapestructs.push_back(tapeStructIth);
diff --git a/mlir/test/Gradient/ConversionTest.mlir b/mlir/test/Gradient/ConversionTest.mlir
index 222a61b5..e1aa2c26 100644
--- a/mlir/test/Gradient/ConversionTest.mlir
+++ b/mlir/test/Gradient/ConversionTest.mlir
@@ -92,7 +92,7 @@ module @test1 {
   }
 
   // CHECK-LABEL: func.func private @rev.rev
-  // CHECK: [[in0ptr:%.+]]: !llvm.ptr, [[diff0ptr:%.+]]: !llvm.ptr, [[out0ptr:%.+]]: !llvm.ptr, [[cotan0ptr:%.+]]: !llvm.ptr, [[tape:%.+]]: !llvm.struct<(struct<(struct<(ptr, ptr, i64)>)>)>) attributes {passthrough = ["noinline"]}
+  // CHECK: [[in0ptr:%.+]]: !llvm.ptr, [[diff0ptr:%.+]]: !llvm.ptr, [[out0ptr:%.+]]: !llvm.ptr, [[cotan0ptr:%.+]]: !llvm.ptr, [[tape:%.+]]: !llvm.struct<(struct<(ptr, ptr, i64)>)>) attributes {passthrough = ["noinline"]}
   gradient.reverse @rev.rev(%arg0: memref<f64>, %arg1: memref<f64>, %arg2: memref<f64>, %arg3: memref<f64>, %arg4: memref<f64>) attributes {argc = 1 : i64, implementation = @rev, resc = 1 : i64, tape = 1 : i64} {
 
     // CHECK: [[in0struct:%.+]] = llvm.load [[in0ptr]] : !llvm.ptr -> !llvm.struct<(ptr, ptr, i64)>
@@ -103,7 +103,7 @@ module @test1 {
     // CHECK: [[out0memref:%.+]] = builtin.unrealized_conversion_cast [[out0struct]] : !llvm.struct<(ptr, ptr, i64)> to memref<f64>
     // CHECK: [[cotan0struct:%.+]] = llvm.load [[cotan0ptr]] : !llvm.ptr -> !llvm.struct<(ptr, ptr, i64)>
     // CHECK: [[cotan0memref:%.+]] = builtin.unrealized_conversion_cast [[cotan0struct]] : !llvm.struct<(ptr, ptr, i64)> to memref<f64>
-    // CHECK: [[tape0struct:%.+]] = llvm.extractvalue [[tape]][0, 0] : !llvm.struct<(struct<(struct<(ptr, ptr, i64)>)>)>
+    // CHECK: [[tape0struct:%.+]] = llvm.extractvalue [[tape]][0] : !llvm.struct<(struct<(ptr, ptr, i64)>)>
     // CHECK: [[tape0memref:%.+]] = builtin.unrealized_conversion_cast [[tape0struct]] : !llvm.struct<(ptr, ptr, i64)> to memref<f64>
     // CHECK: [[result0memref:%.+]] = call @rev([[tape0memref]], [[cotan0memref]])
     %0 = func.call @rev(%arg4, %arg3) : (memref<f64>, memref<f64>) -> memref<f64>
```

The test succeeds, but when running, Enzyme does issue an error:

```python
import pennylane as qml

from catalyst import debug, grad, pure_callback, jacobian

import jax
#x = jax.numpy.array([[3.0, 2.0], [1.0, 1.0]])

@pure_callback
def identity(x) -> float:
    return x


@identity.fwd
def fwd(x):
    return identity(x), (0.5, 0.5) # more than 1


@identity.bwd
def bwd(res, cot):
    return cot * (res[0] + res[1])


@qml.qjit(keep_intermediate=True)
@jacobian
def wrapper(x):
    return identity(x)

x = 1.0
print(wrapper(x))
```

```
 trying to call bwd.rev void (ptr, ptr, ptr, ptr, { { ptr, ptr, i64 }, { ptr, ptr, i64 } })
args[0] =   %14 = alloca { ptr, ptr, i64 }, align 8 FT:ptr
args[1] =   %"'ipa" = alloca { ptr, ptr, i64 }, align 8 FT:ptr
args[2] =   %15 = alloca { ptr, ptr, i64 }, align 8 FT:ptr
args[3] =   %"'ipa3" = alloca { ptr, ptr, i64 }, align 8 FT:ptr
args[4] =   %_augmented = call { { { ptr, ptr, i64 }, { ptr, ptr, i64 } } } @fwd.fwd(ptr %14, ptr %"'ipa", ptr %15, ptr %"'ipa3") FT:{ { ptr, ptr, i64 }, { ptr, ptr, i64 } }
```

But without it, it works alright. With an arbitrary number of elements in the tape.

Reply:
The only reason I'm concerned is that we don't understand the convention correctly then, at least from what we discussed yesterday?

Btw I don't know if this is case, but whenever we said yesterday "so and so is because that's what enzyme expects", it would great to explicitly note down the assumption that we're making there in comments/docstrings. So here as well we could use a comment "Enzyme expects the following format for ... : <...>"

Reply:
I think the issue is that it is also poorly documented on Enzyme. I have tested this experimentally and it works. I'll add the assumption and try to contact the Enzyme authors and submit a PR to Enzyme next quarter. We can make it a ticket :)

Reply:
https://github.com/PennyLaneAI/catalyst/pull/822/commits/fa0f276ed784dd56a8440aea41d52da97d93ccd1

Reply:
So just to confirm, are you saying that in that example enzyme tried to call the backwards with with the results of the forward pass, on top of the residuals?

Looking at this line I can't really tell, that part looks correct no?
```
trying to call bwd.rev void (ptr, ptr, ptr, ptr, { { ptr, ptr, i64 }, { ptr, ptr, i64 } })
```

Reply:
With the change above, I set the tape of the reverse pass to the one given by the message.

```
trying to call bwd.rev void (ptr, ptr, ptr, ptr, { { ptr, ptr, i64 }, { ptr, ptr, i64 } })
```

But if we look at what enzyme is actually passing

```
args[4] =   %_augmented = call { { { ptr, ptr, i64 }, { ptr, ptr, i64 } } } @fwd.fwd(ptr %14, ptr %"'ipa", ptr %15, ptr %"'ipa3")
```

If we look at the Enzyme source:

```
            llvm::errs() << "args[" << i << "] = " << *pre_args[i]
                         << " FT:" << *FT->getParamType(i) << "\n";
```

We see that the expected value of the parameter is

```
FT:{ { ptr, ptr, i64 }, { ptr, ptr, i64 } }
```

But it differs from the argument we are sending. In other words:

sending ` { { { ptr, ptr, i64 }, { ptr, ptr, i64 } } } ` is what enzyme expects

Reply:
Thanks Erick, that's really helpful!